SIWA: Add support for the redirect flow

Tug opened the pull request:
~Requires server patch D33371~ (deployed)

#### Changes proposed in this Pull Request

* Add support for the redirect flow for Sign In with Apple
* Switch to using the redirect flow on `wpcalypso` and `development`, maybe we want `staging` as well?

Using the redirect flow is required so our login is compatible with Safari on iOS 13 and macOS Catalina which bypass the oauth2 flow with a native UI. It seems this UI is not compatible with the popup flow for some unknown reason.

#### Testing instructions
If testing locally, you will need to use the wordpress.com hostname with HTTPS (need a valid redirect uri). You can use Charles proxy to do this (see FG documentation on the subject). Redirect all routes from `https://wordpress.com` to `http://calypso.localhost:3000` except for `/wp-login.php*` which needs to hit the wpcom server (`wordpress.com` or `wpcalypso.wordpress.com` or your sandbox url).

* Try signing in/up with the redirect flow
* Try connecting and disconnecting from Me > Security > Social Login
* Try adding 2FA to your account and sign in
* Try signing in from https://crowdsignal.com
* Try signing in from https://jetpack.com
* Try signing in from https://woocommerce.com
* Try signing in to a wordpress instance with Jetpack using your WPCOM account (Make sure "Allow users to log in to this site using WordPress.com accounts" is enabled in jetpack settings)
* Try linking your Jetpack site to WordPress.com
* Try signing up (if possible for all the flows above) with and without an existing email address in wpcom
* Try a signup flow where `user` is not the first step: Go to http://wordpress.com/themes while logged out, pick a theme, select "Pick this design", then pick a site address then a plan (free for each should do), then click on "Continue with Apple". After authenticating with apple you should come back to the `user` step and it should create your site and log you in
* Make sure we don't break production: set `sign-in-with-apple/redirect` to `false` in `config/development.json` and try signing in/up with Apple, it should work the same as on production.
* Optional: try with Safari on macos Catalina
* Optional: try with Safari on iOS 13 with Charles proxy

Once tested locally let's merge so we can do more tests on wpcalypso. Note that `calypso.live` won't work as the hostname is dynamic which is not supported as a redirect uri for external auth services.

TODO:
- [ ] Make sure we resume the `user` step in the signup flow for flows that don't have user as the first step. For instance starting from https://wordpress.com/start/with-theme

Tug committed:
SIWA: Add support for the redirect flow

Tug committed:
Enable siwa redirect in wpcalypso behind a feature flag

Tug committed:
Make redirectUri optional and give it to the social button when available

Tug committed:
Hide id_token from the callback url

Tug commented on client/components/social-buttons/apple.js:
```diff
@@ -21,25 +22,99 @@ import requestExternalAccess from 'lib/sharing';
+					clientId: this.props.clientId,
+					scope: this.props.scope,
+					redirectURI: this.props.redirectUri,
+					state: '1',
```
Need to see if we need to set a state to prevent CSRF

Tug replied:
Here https://auth0.com/docs/protocols/oauth2/mitigate-csrf-attacks they recommend using the localStorage for the state in the case of a SPA. I might just do that 👍

sgomes commented on client/blocks/login/social.jsx:
```diff
@@ -166,14 +174,14 @@ class SocialLoginForm extends Component {
-					login( { isNative: true, socialService: service } ) }`
-			: null;
+		const host = typeof window !== 'undefined' && window.location.host;
+		return `https://${ host + login( { isNative: true, socialService: service } ) }`;
```
This appears to ignore `uxMode` now, for every service (not just Apple). Is this intentional? If so, and if it's not needed for anything, we should drop the parameter to avoid confusion.

Tug replied:
Yes, I wanted to give the redirect uri to both login buttons in all cases, even when the uxMode is `popup`. We might need it anyway to initialize the client (the apple client for instance will throw if we call init without a redirect uri).

sgomes commented on client/blocks/login/social.jsx:
```diff
@@ -108,12 +108,20 @@ class SocialLoginForm extends Component {
+			redirectTo = window.sessionStorage.getItem( 'login_redirect_to' );
+		}
+
+		window.sessionStorage.removeItem( 'login_redirect_to' );
```
Can `handleAppleResponse` ever happen on the server? I'm assuming it can't, but otherwise this line (and the similar one above) would need to be guarded.

Tug replied:
Not sure what you mean by guarded here.
But yeah in theory it would be better to handle the redirect on the server, we actually already do so for the popup flow but it's limited to a single redirect uri while we need to handle many flows and redirects on calypso.

sgomes replied:
I meant doing something along the lines of `typeof window !== 'undefined' && window.sessionStorage...`. If this won't run on the server, though, that's not needed.

Tug replied:
Oh right 👍 It shouldn't run on the server but I'll double check!

sgomes commented on client/blocks/login/social.jsx:
```diff
@@ -182,12 +190,21 @@ class SocialLoginForm extends Component {
 						responseHandler={ this.handleAppleResponse }
-						onClick={ this.trackLogin.bind( null, 'apple' ) }
+						uxMode={ uxModeApple }
+						redirectUri={ this.getRedirectUrl( uxMode, 'apple' ) }
```
If the `uxMode` parameter really is needed, make sure you really want to pass `uxMode` and not `uxModeApple` here.

Tug replied:
Indeed `uxMode` is not needed anymore!

sgomes commented on client/blocks/login/social.jsx:
```diff
@@ -182,12 +190,21 @@ class SocialLoginForm extends Component {
-						onClick={ this.trackLogin.bind( null, 'google' ) }
+						onClick={ this.trackLoginAndRememberRedirect.bind( null, 'google' ) }
+						socialServiceResponse={
+							this.props.socialService === 'google' ? this.props.socialServiceResponse : null
```
As a suggestion for a future improvement, I would consider rearchitecting this mechanism. From what I'm reading here, it appears to me that this component can receive both Google and Apple responses as a prop, and needs to look at which one it's getting in order to populate the correct subcomponent's props?

That seems to lead to a lot of special casing everywhere, so perhaps it would make sense to extract this logic somehow. Perhaps add an Apple-specific and Google-specific layer in between this and the buttons? Sorry I can't provide a more detailed alternative, but I'm still trying to grok the whole thing 😕

Tug replied:
Agreed it's not the cleanest thing and could benefit a refactor 👍

sgomes commented on client/blocks/signup-form/social.jsx:
```diff
@@ -81,7 +81,9 @@ class SocialSignupForm extends Component {
 	render() {
 		const uxMode = this.shouldUseRedirectFlow() ? 'redirect' : 'popup';
-		const redirectUri = uxMode === 'redirect' ? `https://${ window.location.host }/start` : null;
+		const host = typeof window !== 'undefined' && window.location.host;
```
Most of these changes are duplicate logic from the file above. It would be really nice to extract it somehow 😕

Tug replied:
Agreed there is a some repetition here, we're not generating the same redirectUri though so I'm not sure it's worth it

sgomes replied:
Ah, I missed that difference, thanks!

sgomes reviewed (commented):
Thank you for the great PR, @Tug! 

I reviewed your code from a very high level, since I'm missing a lot of detail on how all of these pieces play together. I added some notes and ideas for potential improvements, either as part of this PR or a future one.

Also, your code seems to play nicely with the login entrypoint change that was merged earlier today, so there shouldn't be any issues on that front! 👍

Tug committed:
Add OAuth2 state param

Tug committed:
Remove unused parameter in getRedirectUrl

Tug committed:
Use async/await syntax instead of promises

Tug committed:
Rename appleClientLoaded to appleClient as it references the Apple lib

Tug committed:
Add support from SIWA redirect in Me > Security > Social Login

Tug commented on client/me/social-login/action-button.jsx:
```diff
@@ -119,8 +120,16 @@ class SocialLoginActionButton extends Component {
 
 		if ( service === 'apple' ) {
+			const uxMode = config.isEnabled( 'sign-in-with-apple/redirect' ) ? 'redirect' : 'popup';
+			const redirectUri = window.location.origin + window.location.pathname;
```
here we should add safeguards though

Tug committed:
Add safeguards for window in SocialLoginActionButton

belcherj commented on client/login/controller.js:
```diff
@@ -27,13 +27,19 @@ const enhanceContextWithLogin = context => {
+	const previousHash = context.state || {};
+	const { client_id, user_email, user_name, id_token, state } = previousHash;
+	const socialServiceResponse = client_id
+		? { client_id, user_email, user_name, id_token, state }
```
why not just use `previousHash`

Tug replied:
Because previousHash might contain more properties that we don't want in our `socialServiceResponse`. Moreover, destructuring `previousHash` makes sure you'll get each property in `socialServiceResponse` even if it's not present in the hash object

belcherj commented on client/me/security/controller.js:
```diff
@@ -62,9 +63,24 @@ export function accountRecovery( context, next ) {
+	}
+
+	const previousHash = context.state || {};
+	const { client_id, user_email, user_name, id_token, state } = previousHash;
```
doesn't seem necessary to destructure here

Tug replied:
Same reason I think it's a nice way to make sure our response has all properties, no more no less (also useful to document what we expect the `socialServiceResponse` format to be.

belcherj reviewed (approved):
Just a couple small syntactical issues.

Tug commented on client/components/social-buttons/apple.js:
```diff
@@ -21,25 +22,101 @@ import requestExternalAccess from 'lib/sharing';
 
+	appleClient = null;
+
+	componentDidMount() {
```
Not sure what's wrong with `componentDidMount`? If it isn't allowed anymore I'm not aware of it

belcherj replied:
https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html

Tug commented on client/components/social-buttons/apple.js:
```diff
@@ -21,25 +22,101 @@ import requestExternalAccess from 'lib/sharing';
+			return;
+		}
+
+		if ( this.props.uxMode === 'redirect' ) {
```
Sure I can extract that into a separate function 👍

Tug committed:
Refactor handling socialServiceResponse in componentDidMount

Tug committed:
Make sure we don't send a boolean for socialServiceResponse

Tug committed:
Resume previous flow when coming back from an oauth2 redirect

Tug commented:
@sgomes @belcherj Thanks for reviewing. I've addressed your feedback and just added one more thing: a fix (in https://github.com/Automattic/wp-calypso/pull/36444/commits/042f8ebfd5ed023bafdd6d751bb1366e80aaf6c1) to resume a signup flow when we come back from apple. Without this, the user will be redirected to the default flow (onboarding) and will have to start over all the steps. I've added a test case in the PR description for this.

astralbodies commented:
I tried testing with my vm of Catalina and I think I'm running into limitations of the local dev test environment. 

After logging in with the native UI (it did come up!) it redirected over to `https://wordpress.com/log-in/apple/callback` and shows the login screen again. In the JavaScript console I see a 404 for `https://wordpress.com/wp-login.php?action=social-login-endpoint`. I think this might be due to nginx intercepting some of these links on the actual .com side and it's not happening locally.

astralbodies reviewed (approved):
I tested this as best as I could on macOS Catalina with Charles Proxy doing a bunch of the work. There were glitches in Firefox but given the 404s I was seeing I think it has something to do with my rewrite setup, not this PR. I'm going to give it a 👍 for now because I don't know how much else I can functionally test it.

belcherj committed:
Add flag for staging
